### ISSUE TITLE
Add pyright type-checking

### ISSUE THREAD

**Issue by MartinBernstorff** [marimo-team/marimo#1163]
### Description

Love the inline type-checking in VSCode, and would love to see it here as well. 

### Suggested solution

Add pyright type-checking and underlining.

**Comment by akshayka** [comment]
Thanks for the suggestion. We'd like to add a language server at some point, and this would be a great feature.

**Comment by aszenz** [comment]
I don't even see syntax error checking in the editor, this makes it difficult to work in.
So basic diagnostics and linting would be welcome.

**Comment by mscolnick** [comment]
@aszenz, this is coming in #3414

**Comment by MartinBernstorff** [comment]
@mscolnick Apologies, wanted to unsubscribe, not close :-)

**Comment by mscolnick** [comment]
for the folks interested in `pyright`. we do have type-checking via mypy and python-lsp:

```
pip install "marimo[lsp]" pylsp-mypy
```

And enable `Diagnostics` in the settings modal

![Image](https://github.com/user-attachments/assets/3420aa6c-2f9e-4af3-80a7-5d3bd9276b68)

I am curious if this works for folks, or if they are specifically looking for pyright (and not mypy)

**Comment by ilyagr** [comment]
There are now two brand new typecheckers that claim to be quite fast, both written in Rust:

Pyrefly, https://pyrefly.org/ (already has IDE support)

ty, https://github.com/astral-sh/ty (shares some code with `ruff`, doesn't claim to be "production ready" yet).

I imagine both try to improve on top of pyright. If you scroll down a bit on Pyrefly's website I linked, it claims to be more than 10x faster than pyright or mypy.

**Comment by winstxnhdw** [comment]
I am interested in replacing Pyright with ty in marimo. Will experiment on some POCs and maybe a draft PR, then we can later evaluate if we would like to be an early adopter.

As of now, ty is more performant than Pyrefly in latest benchmarks and I am betting that ty will remain the winner in this race, for both strictness and performance.

**Comment by ilyagr** [comment]
There's an interesting comparison of pyrefly and ty: <https://blog.edward-li.com/tech/comparing-pyrefly-vs-ty/>, <https://lobste.rs/s/2uukup/pyrefly_vs_ty_comparing_python_s_two_new>.

TLDR: they have quite different philosophies in terms of whether to prioritize correctness or ease of use to gradually add types to an existing codebase.

So, it might take some thought to decide which works better for marimo, and it might make sense to support both.

**Comment by mscolnick** [comment]
my observation (and personal anecdote) of notebooks is that people rarely use types and if so, they use them sparingly and heavy on inference →  so i think prioritizing a type-checker that has good inference and allows for "gradually adding types" does seem the best to me

**Comment by firai** [comment]
To add to the list of Python type-checker/LSP options, [`basedpyright`](https://github.com/DetachHead/basedpyright), which is a fork of `pyright`, is a balance between the mature (but somewhat slow) `python-lsp-server` and the speed (but not very mature) `ty`/`pyrefly`.

**Comment by superlopuh** [comment]
I'd just like to add that our project (https://github.com/xdslproject/xdsl) relies on Pyright's experimental [TypeForm](https://peps.python.org/pep-0747/) support, so we would really like Pyright specifically, and hopefully in a way that can see our pyproject.toml settings.

**Comment by winstxnhdw** [comment]
> I'd just like to add that our project (https://github.com/xdslproject/xdsl) relies on Pyright's experimental [TypeForm](https://peps.python.org/pep-0747/) support, so we would really like Pyright specifically, and hopefully in a way that can see our pyproject.toml settings.

No worries. Once `ty` integration is done, it'll be trivial to add `pyright`. I am working on it today.

**Comment by ShravanSunder** [comment]
pyright please! :)

**Comment by z0u** [comment]
I would like `basedpyright` (specifically): unlike `pyright`, it gives exactly the same errors when running in CI as it does in the editor. I found that there was no way to resolve the differences between `pyright` as a language server, and `pyright` as a command — whereas `basedpyright` just works. I also tried `ty`, but it gives too many false positives due to lacking features.

I do use types in my Jupyter notebooks, and I'm hoping to migrate to marimo.

Examples of my use of types:
- [Small notebook: Modal demo](https://github.com/z0u/mi-ni/blob/c0a0134cef2536545033f9423a6084fe8097f7f7/docs/getting-started.ipynb)
- [Large notebook: training an MLP](https://github.com/z0u/ex-color-transformer/blob/c7769cf1f950595b549141e422f9de5c096a4c66/docs/m1-color-mlp/ex-1.7-sparse-labels.ipynb)
- [pyproject.toml](https://github.com/z0u/mi-ni/blob/c0a0134cef2536545033f9423a6084fe8097f7f7/pyproject.toml#L117)
- [CI run](https://github.com/z0u/mi-ni/actions/runs/15845599144/job/44666919398) (see _Type checking_ step)
